src/DateRangePicker.ts

```diff
@@ -1,26 +1,491 @@
+  TemplateResult,
+  SVGTemplateResult,
+  property,
+  query,
+} from 'lit-element';
+import dayjs from 'dayjs/esm/index.js';
+
+// these values can be overridden via the component's HTML (camelCased) attributes
+const WIDTH = 180;
+const HEIGHT = 40;
+const SLIDER_WIDTH = 10;
+const TOOLTIP_WIDTH = 125;
+const TOOLTIP_HEIGHT = 30;
+const DATE_FORMAT = 'M/D/YYYY';
+
+// this constant is not set up to be overridden
+const SLIDER_CORNER_SIZE = 4;
+
+// these CSS custom props can be overridden from the HTML that is invoking this component
+const sliderFill = 'var(--dateRangePickerSliderFill, #4B65FE)';
+const selectedRangeFill = 'var(--dateRangePickerSelectedRangeFill, #DBE0FF)';
+const barIncludedFill = 'var(--dateRangePickerBarIncludedFill, #2C2C2C)';
+const barExcludedFill = 'var(--dateRangePickerBarExcludedFill, #CCCCCC)';
+const inputBorder = css`var(--dateRangePickerInputBorder, 0.5px solid #2C2C2C)`;
+const inputWidth = css`var(--dateRangePickerInputWidth, 70px)`;
+const inputFontSize = css`var(--dateRangePickerInputFontSize, 1.2rem)`;
+const tooltipBackgroundColor = css`var(--dateRangePickerTooltipBackgroundColor, #2C2C2C)`;
+const tooltipTextColor = css`var(--dateRangePickerTooltipTextColor, #FFFFFF)`;
+const tooltipFontSize = css`var(--dateRangePickerTooltipFontSize, 1.1rem)`;
+
+type SliderIds = 'slider-min' | 'slider-max';
+
+interface HistogramInputData {
+  minDate: string;
+  maxDate: string;
+  bins: number[];
+}
+
+interface HistogramItem {
+  value: number;
+  height: number;
+  binStart: string;
+  binEnd: string;
+}
 
 export class DateRangePicker extends LitElement {
-  static styles = css`
-    :host {
-      display: block;
-      padding: 25px;
-      color: var(--date-range-picker-text-color, #000);
+  /* eslint-disable lines-between-class-members */
+
+  // these properties are intended to be passed in as attributes
+  @property({ type: Number }) width = WIDTH;
+  @property({ type: Number }) height = HEIGHT;
+  @property({ type: Number }) sliderWidth = SLIDER_WIDTH;
+  @property({ type: Number }) tooltipWidth = TOOLTIP_WIDTH;
+  @property({ type: Number }) tooltipHeight = TOOLTIP_HEIGHT;
+  @property({ type: String }) dateFormat = DATE_FORMAT;
+  @property({ type: Object }) data?: HistogramInputData;
+
+  @internalProperty() minSliderX: number = 0;
+  @internalProperty() maxSliderX: number = 0;
+  @internalProperty() tooltipOffset = 0;
+  @internalProperty() tooltipContent: TemplateResult | '' = '';
+  @internalProperty() tooltipDisplay: 'block' | 'none' = 'none';
+
+  @query('#tooltip') tooltip!: HTMLDivElement;
+  @query('#container') container!: HTMLDivElement;
+
+  // these properties don't need to be tracked for changes
+  private _minDate: number = 0;
+  private _maxDate: number = 0;
+  private _dragOffset: number = 0;
+  private _histWidth: number = 0;
+  private _numBins: number = 0;
+  private _binWidth: number = 0;
+  private _currentSlider?: SVGRectElement;
+  private _histData: HistogramItem[] = [];
```

Comment:
`_histData` should be reactive so the UI updates when it updates. The general idea with components is you update the properties, all of the dependencies of that property should update. Right now, you're only updating `_histData` on `firstUpdated` so any subsequent updates to `data` won't change the histogram.

Reply:
I just pushed a quick change that makes the data property reactive. ie you can now change the data on-the-fly and it will update the component in real time: 

```
document.querySelector('histogram-date-range').setAttribute('data', '{ "minDate": "May 1, 1972", "maxDate": "12/21/1981","bins": [ 85, 25, 56, 1000, 12, 101]}')

```
I understand this is probably still not ideal reactive design but what do you think of this as an intermediate step (so I can keep on trucking towards getting a first version of this deployed)?

Reply:
Cool looks good 👍